File: deployment-package/proxy.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
微信API代理服务器 - Python/Flask 版本
用于转发微信公众号请求到 Supabase Edge Functions
"""

import os
import time
import requests
from flask import Flask, request, jsonify, Response
from flask_cors import CORS
from datetime import datetime

app = Flask(__name__)
CORS(app)

# 配置
PORT = int(os.getenv('PORT', 3000))
PROXY_AUTH_TOKEN = os.getenv('PROXY_AUTH_TOKEN')
EDGE_FUNCTION_URL = os.getenv('EDGE_FUNCTION_URL', '')

# 日志函数
def log(message):
    timestamp = datetime.now().isoformat()
    print(f'[{timestamp}] {message}')

# 请求日志中间件
@app.before_request
def log_request():
    log(f'{request.method} {request.path}')

# 健康检查端点
@app.route('/health', methods=['GET'])
def health():
    return jsonify({
        'status': 'ok',
        'timestamp': datetime.now().isoformat(),
        'server': 'WeChat API Proxy (Python/Flask)',
        'version': '1.0.0',
        'uptime': time.process_time()
    })

# 微信回调端点 - GET (URL验证)
@app.route('/wechat-callback', methods=['GET'])
def wechat_callback_get():
    try:
        # 获取微信验证参数
        signature = request.args.get('signature', '')
        timestamp = request.args.get('timestamp', '')
        nonce = request.args.get('nonce', '')
        echostr = request.args.get('echostr', '')
        
        log(f'微信URL验证请求: signature={signature}, timestamp={timestamp}, nonce={nonce}')
        
        if not EDGE_FUNCTION_URL:
            log('错误: EDGE_FUNCTION_URL 未配置')
            return jsonify({'error': 'EDGE_FUNCTION_URL not configured'}), 500
        
        # 转发到 Edge Function
        target_url = f'{EDGE_FUNCTION_URL}?signature={signature}&timestamp={timestamp}&nonce={nonce}&echostr={echostr}'
        log(f'转发验证请求到: {target_url}')
        
        response = requests.get(target_url, timeout=10)
        log(f'Edge Function 响应状态: {response.status_code}')
        
        # 返回 echostr 完成验证
        return Response(response.text, status=response.status_code, mimetype='text/plain')
        
    except Exception as e:
        log(f'URL验证错误: {str(e)}')
        return jsonify({'error': f'URL validation failed: {str(e)}'}), 500

# 微信回调端点 - POST (接收消息)
@app.route('/wechat-callback', methods=['POST'])
def wechat_callback_post():
    try:
        # 获取请求参数
        signature = request.args.get('signature', '')
        timestamp = request.args.get('timestamp', '')
        nonce = request.args.get('nonce', '')
        openid = request.args.get('openid', '')
        encrypt_type = request.args.get('encrypt_type', '')
        msg_signature = request.args.get('msg_signature', '')
        
        # 获取原始请求体（XML）
        body = request.get_data(as_text=True)
        
        log(f'接收微信消息: openid={openid}, encrypt_type={encrypt_type}')
        
        if not EDGE_FUNCTION_URL:
            log('错误: EDGE_FUNCTION_URL 未配置')
            return jsonify({'error': 'EDGE_FUNCTION_URL not configured'}), 500
        
        # 构建转发URL
        params = []
        if signature:
            params.append(f'signature={signature}')
        if timestamp:
            params.append(f'timestamp={timestamp}')
        if nonce:
            params.append(f'nonce={nonce}')
        if openid:
            params.append(f'openid={openid}')
        if encrypt_type:
            params.append(f'encrypt_type={encrypt_type}')
        if msg_signature:
            params.append(f'msg_signature={msg_signature}')
        
        target_url = f'{EDGE_FUNCTION_URL}?{"&".join(params)}'
        log(f'转发消息到: {target_url}')
        
        # 转发请求
        headers = {
            'Content-Type': 'application/xml',
        }
        
        response = requests.post(
            target_url,
            data=body.encode('utf-8'),
            headers=headers,
            timeout=30
        )
        
        log(f'Edge Function 响应状态: {response.status_code}')
        
        # 返回 Edge Function 的响应
        return Response(
            response.content,
            status=response.status_code,
            mimetype='application/xml'
        )
        
    except Exception as e:
        log(f'消息处理错误: {str(e)}')
        return jsonify({'error': f'Message processing failed: {str(e)}'}), 500

# 微信API代理端点
@app.route('/wechat-proxy', methods=['POST'])
def wechat_proxy():
    try:
        # 验证认证令牌
        if PROXY_AUTH_TOKEN:
            auth_header = request.headers.get('Authorization', '')
            token = auth_header.replace('Bearer ', '')
            
            if token != PROXY_AUTH_TOKEN:
                log('未授权访问尝试')
                return jsonify({'error': 'Unauthorized'}), 401
        
        data = request.json
        target_url = data.get('target_url')
        method = data.get('method', 'GET')
        headers = data.get('headers', {})
        body = data.get('body')
        
        if not target_url:
            return jsonify({'error': 'target_url is required'}), 400
        
        log(f'代理请求 {method} {target_url}')
        
        # 准备请求
        request_headers = {
            'Content-Type': 'application/json',
            **headers
        }
        
        # 发送请求
        if method.upper() in ['POST', 'PUT', 'PATCH']:
            response = requests.request(
                method,
                target_url,
                json=body,
                headers=request_headers,
                timeout=30
            )
        else:
            response = requests.request(
                method,
                target_url,
                headers=request_headers,
                timeout=30
            )
        
        log(f'响应状态: {response.status_code}')
        
        # 返回响应
        try:
            return jsonify(response.json()), response.status_code
        except:
            return Response(response.text, status=response.status_code)
            
    except Exception as e:
        log(f'代理错误: {str(e)}')
        return jsonify({
            'error': 'Proxy request failed',
            'message': str(e)
        }), 500

# 404 处理
@app.errorhandler(404)
def not_found(e):
    return jsonify({'error': 'Not found'}), 404

# 错误处理
@app.errorhandler(Exception)
def handle_error(e):
    log(f'错误: {str(e)}')
    return jsonify({'error': 'Internal server error'}), 500

# 启动服务器
if __name__ == '__main__':
    print('=' * 60)
    print('微信API代理服务器 (Python/Flask)')
    print('=' * 60)
    print(f'状态: 运行在端口 {PORT}')
    print(f'健康检查: http://localhost:{PORT}/health')
    print(f'代理端点: http://localhost:{PORT}/wechat-proxy')
    print(f'微信回调: http://localhost:{PORT}/wechat-callback')
    print(f'认证: {"已启用" if PROXY_AUTH_TOKEN else "已禁用 (⚠️  不安全!)"}')
    print(f'Edge Function: {EDGE_FUNCTION_URL or "未配置"}')
    print(f'启动时间: {datetime.now().isoformat()}')
    print('=' * 60)
    
    # 生产环境使用 gunicorn，开发环境使用 Flask 内置服务器
    app.run(host='0.0.0.0', port=PORT, debug=False)

File: deployment-package/proxy.js
const express = require('express');
const cors = require('cors');

const app = express();
const PORT = process.env.PORT || 3000;
const PROXY_AUTH_TOKEN = process.env.PROXY_AUTH_TOKEN;

// 🔒 SECURITY: Mandatory authentication - server refuses to start without token
if (!PROXY_AUTH_TOKEN) {
  console.error('='.repeat(60));
  console.error('❌ SECURITY ERROR: PROXY_AUTH_TOKEN is required');
  console.error('='.repeat(60));
  console.error('The WeChat API proxy server cannot start without authentication.');
  console.error('');
  console.error('To fix this:');
  console.error('  1. Generate a secure token: openssl rand -hex 32');
  console.error('  2. Set environment variable: export PROXY_AUTH_TOKEN=<your-token>');
  console.error('  3. Or add to .env file: PROXY_AUTH_TOKEN=<your-token>');
  console.error('');
  console.error('See deployment-package/README.md for more details.');
  console.error('='.repeat(60));
  process.exit(1);
}

// 🔒 SECURITY: Validate token strength (minimum 32 characters)
if (PROXY_AUTH_TOKEN.length < 32) {
  console.error('='.repeat(60));
  console.error('❌ SECURITY ERROR: PROXY_AUTH_TOKEN is too weak');
  console.error('='.repeat(60));
  console.error('Token must be at least 32 characters for security.');
  console.error('');
  console.error('Generate a secure token: openssl rand -hex 32');
  console.error('='.repeat(60));
  process.exit(1);
}

// 启用CORS和JSON解析
app.use(cors());
app.use(express.json({ limit: '10mb' }));

// 请求日志中间件
app.use((req, res, next) => {
  const timestamp = new Date().toISOString();
  console.log(`[${timestamp}] ${req.method} ${req.path}`);
  next();
});

// 健康检查端点
app.get('/health', (req, res) => {
  res.json({ 
    status: 'ok', 
    timestamp: new Date().toISOString(),
    server: 'WeChat API Proxy',
    version: '1.0.0',
    uptime: process.uptime()
  });
});

// 🔒 Authentication middleware - always enforced (token is mandatory at startup)
const authenticateRequest = (req, res, next) => {
  const authHeader = req.headers.authorization;
  const token = authHeader?.replace('Bearer ', '');
  
  if (token !== PROXY_AUTH_TOKEN) {
    console.error(`[${new Date().toISOString()}] Unauthorized access attempt from ${req.ip}`);
    return res.status(401).json({ error: 'Unauthorized' });
  }
  next();
};

// 微信access_token获取端点
app.post('/wechat/token', authenticateRequest, async (req, res) => {
  try {
    const { appid, secret } = req.body;

    if (!appid || !secret) {
      return res.status(400).json({ error: 'appid and secret are required' });
    }

    const targetUrl = `https://api.weixin.qq.com/cgi-bin/token?grant_type=client_credential&appid=${appid}&secret=${secret}`;
    console.log(`[${new Date().toISOString()}] 获取微信access_token for appid: ${appid}`);

    const response = await fetch(targetUrl);
    const data = await response.json();

    console.log(`[${new Date().toISOString()}] 微信API响应: ${data.access_token ? '成功' : '失败 - ' + data.errmsg}`);

    res.status(response.status).json(data);
  } catch (error) {
    console.error(`[${new Date().toISOString()}] Token获取失败:`, error.message);
    res.status(500).json({ 
      error: 'Token request failed', 
      message: error.message 
    });
  }
});

// 微信二维码创建端点
app.post('/wechat/qrcode/create', authenticateRequest, async (req, res) => {
  try {
    const { access_token, expire_seconds, action_name, action_info } = req.body;

    if (!access_token) {
      return res.status(400).json({ error: 'access_token is required' });
    }

    const targetUrl = `https://api.weixin.qq.com/cgi-bin/qrcode/create?access_token=${access_token}`;
    console.log(`[${new Date().toISOString()}] 创建微信二维码, action_name: ${action_name}`);

    const response = await fetch(targetUrl, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ expire_seconds, action_name, action_info }),
    });

    const data = await response.json();

    console.log(`[${new Date().toISOString()}] 二维码创建: ${data.ticket ? '成功' : '失败 - ' + data.errmsg}`);

    res.status(response.status).json(data);
  } catch (error) {
    console.error(`[${new Date().toISOString()}] 二维码创建失败:`, error.message);
    res.status(500).json({ 
      error: 'QR code creation failed', 
      message: error.message 
    });
  }
});

// 微信API代理端点
app.post('/wechat-proxy', authenticateRequest, async (req, res) => {
  try {
    const { target_url, method = 'GET', headers = {}, body } = req.body;

    if (!target_url) {
      return res.status(400).json({ error: 'target_url is required' });
    }

    console.log(`[${new Date().toISOString()}] Proxying ${method} ${target_url}`);

    // 准备请求选项
    const fetchOptions = {
      method: method,
      headers: {
        'Content-Type': 'application/json',
        ...headers,
      },
    };

    // 添加请求体（如果有）
    if (body && (method === 'POST' || method === 'PUT' || method === 'PATCH')) {
      fetchOptions.body = typeof body === 'string' ? body : JSON.stringify(body);
    }

    // 转发请求到微信API
    const response = await fetch(target_url, fetchOptions);
    const data = await response.json();

    console.log(`[${new Date().toISOString()}] Response status: ${response.status}`);

    // 返回微信API的响应
    res.status(response.status).json(data);
  } catch (error) {
    console.error(`[${new Date().toISOString()}] Proxy error:`, error.message);
    res.status(500).json({ 
      error: 'Proxy request failed', 
      message: error.message 
    });
  }
});

// 微信支付回调转发端点（无需认证 - 微信服务器直接调用）
app.post('/wechat-pay-callback', async (req, res) => {
  const timestamp = new Date().toISOString();
  console.log(`[${timestamp}] 📥 微信支付回调收到`);
  
  try {
    // 获取原始请求体
    const body = JSON.stringify(req.body);
    console.log(`[${timestamp}] 回调数据:`, body.substring(0, 200) + '...');
    
    // 转发到 Supabase Edge Function
    const targetUrl = 'https://vlsuzskvykddwrxbmcbu.supabase.co/functions/v1/wechat-pay-callback';
    
    const response = await fetch(targetUrl, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: body,
    });
    
    const data = await response.json();
    console.log(`[${timestamp}] ✅ 回调转发结果:`, JSON.stringify(data));
    
    res.status(response.status).json(data);
  } catch (error) {
    console.error(`[${timestamp}] ❌ 支付回调转发失败:`, error.message);
    res.status(500).json({ 
      code: 'FAIL', 
      message: error.message 
    });
  }
});

// 小程序登录端点（无需认证 - 小程序直接调用）
app.post('/miniprogram-login', async (req, res) => {
  const timestamp = new Date().toISOString();
  console.log(`[${timestamp}] 📱 小程序登录请求`);
  
  try {
    const { code } = req.body;
    
    if (!code) {
      console.error(`[${timestamp}] ❌ 缺少 code 参数`);
      return res.status(400).json({ 
        success: false, 
        error: '缺少 code 参数' 
      });
    }
    
    console.log(`[${timestamp}] 小程序 code: ${code.substring(0, 10)}...`);
    
    // 从环境变量获取小程序配置
    const appId = process.env.WECHAT_MINI_PROGRAM_APP_ID;
    const appSecret = process.env.WECHAT_MINI_PROGRAM_APP_SECRET;
    
    if (!appId || !appSecret) {
      console.error(`[${timestamp}] ❌ 小程序配置不完整`);
      return res.status(500).json({ 
        success: false, 
        error: '小程序配置不完整' 
      });
    }
    
    // 调用微信 jscode2session API
    const wxUrl = `https://api.weixin.qq.com/sns/jscode2session?appid=${appId}&secret=${appSecret}&js_code=${code}&grant_type=authorization_code`;
    
    console.log(`[${timestamp}] 调用微信 jscode2session API...`);
    
    const wxResponse = await fetch(wxUrl);
    const wxResult = await wxResponse.json();
    
    console.log(`[${timestamp}] 微信响应:`, { 
      openid: wxResult.openid ? wxResult.openid.substring(0, 10) + '...' : null,
      errcode: wxResult.errcode,
      errmsg: wxResult.errmsg
    });
    
    if (wxResult.errcode) {
      console.error(`[${timestamp}] ❌ 微信接口错误: ${wxResult.errcode} - ${wxResult.errmsg}`);
      return res.status(400).json({ 
        success: false, 
        error: `微信接口错误: ${wxResult.errcode} - ${wxResult.errmsg}` 
      });
    }
    
    if (!wxResult.openid) {
      console.error(`[${timestamp}] ❌ 未获取到 openid`);
      return res.status(400).json({ 
        success: false, 
        error: '未获取到 openid' 
      });
    }
    
    console.log(`[${timestamp}] ✅ 小程序登录成功`);
    
    // 返回 openid（不返回 session_key，保证安全）
    res.json({
      success: true,
      openid: wxResult.openid,
      unionid: wxResult.unionid || null,
    });
    
  } catch (error) {
    console.error(`[${timestamp}] ❌ 小程序登录失败:`, error.message);
    res.status(500).json({ 
      success: false, 
      error: error.message 
    });
  }
});

// 404 处理
app.use((req, res) => {
  res.status(404).json({ error: 'Not found' });
});

// 错误处理中间件
app.use((err, req, res, next) => {
  console.error(`[${new Date().toISOString()}] Error:`, err);
  res.status(500).json({ error: 'Internal server error' });
});

// 启动服务器
app.listen(PORT, '0.0.0.0', () => {
  console.log('='.repeat(60));
  console.log('🔒 WeChat API Proxy Server (Secured)');
  console.log('='.repeat(60));
  console.log(`Status: Running on port ${PORT}`);
  console.log(`Health check: http://localhost:${PORT}/health`);
  console.log(`Proxy endpoint: http://localhost:${PORT}/wechat-proxy`);
  console.log(`Authentication: ENABLED ✅`);
  console.log(`Token length: ${PROXY_AUTH_TOKEN.length} characters`);
  console.log(`Started at: ${new Date().toISOString()}`);
  console.log('='.repeat(60));
});

// 优雅关闭
process.on('SIGINT', () => {
  console.log('\nShutting down gracefully...');
  process.exit(0);
});

process.on('SIGTERM', () => {
  console.log('\nShutting down gracefully...');
  process.exit(0);
});
